### DISCUSSION TITLE
How to properly end a scan?

### DISCUSSION THREAD

**Discussion by mscottwvc** [google/bumble#542] [Q&A] [answered]
Looking at the run_scanner.py example.  If I wanted to cleanly shutdown the script after finding a specific mac address,  what would be the proper steps be?  Is it as simple as `await device.stop_scanning()` or are there more things I should do?

**Comment by barbibulle** [comment, marked as answer]
`await device.stop_scanning()` is all that's needed.

**Reply by mscottwvc** [reply]
@barbibulle 
Thanks I thought that would be enough for me, but I still seem to be failing.  I should have said I wanted to do a clean shutdown in a signal handler, which I thought would be simple but so far I haven't been completely successful.  I made device global and added a signal handler and inside did: 
`asyncio.create_task(handleStop)`
then my handleStop function:
```
async def handleStop():
  await device.stop_scanning()
```
Scanning stops but the program doesnt end.  I guess I could call exit then, but I would rather be safe and have a clean program shutdown.  I didnt run a debugger yet, but my guess is that something in this line is blocking
`await hci_source.wait_for_termination()`

Could you please give me any pointers?

I also looked the scan app, it looks like its use click and appears to inherit that signal handler (when I do ctrl+c I see Aborted!) so I am guessing that does not shutdown cleanly as is (not that you should change it to have that requirement, just checking my understanding).

**Reply by barbibulle** [reply]
Using async functions in a synchronous callback can be tricky. `asyncio.create_task(handleStop)` looks like it should do what you want, but in reality, you can run into a race condition, because if you don't hold a reference to the task, it may be released before it has a chance to run. That's a known gotcha with python asyncio. Luckily, there are some utility functions to help with this.
One option is to use the `utils.AsyncRunner.spawn(func)` functions (it will create a task for you, keep a reference, and release the reference when the task completes). An alternative option it to use the `utils.AsyncRunner.run_in_task()` decorator for your event handler function (sync function).
Note that `await hci_source.wait_for_termination()` will not return until the transport source is terminated (how that happens depends on the source type: for example, a TCP source will terminate when the connection is closed). 
Handling ctrl-c is quite tricky with asyncio, because when that's caught, the asyncio subsystem automatically cancels all running tasks, so once the handler is invoked, you really can't cleanly make more async calls, so cleaning up becomes difficult. There are ways around that, but that's too complex to use in the simple example apps, that's why they don't bother cleanly exiting on ctrl-c (they just exit, but won're reset the controller for example).

**Reply by mscottwvc** [reply]
Thanks for the detailed explanation!  I will look into the utils and use a non signal handler approach.  

If I have a hardware USB device as my source, what is the recommended way to terminate or close that?

**Reply by zxzxwu** [reply]
USB can usually be released automatically after you exit the script. Or, you can try to explicitly close the transport. There is a close method.
